Projects/Server/Network/RingPipe.cs

```diff
@@ -0,0 +1,271 @@
+            // the read pointer. Check that here.
+            if (write == read)
+            {
+                throw new InvalidOperationException("Write index equals read index after advance");
+            }
+
+            _pipe._writeIdx = write;
+        }
+
+        public void Close() => _pipe._closed = true;
+
+        public bool IsClosed => _pipe._closed;
+    }
+
+    public class PipeReader
+    {
+        private readonly RingPipe _pipe;
+
+        internal PipeReader(RingPipe pipe) => _pipe = pipe;
+
+        public unsafe Span<byte> TryRead()
+        {
+            var read = _pipe._readIdx;
+            var write = _pipe._writeIdx;
+
+            uint sz;
+            if (read <= write)
+            {
+                sz = write - read;
+            }
+            else
+            {
+                sz = _pipe.Size - read + write;
+            }
+
+            return new Span<byte>((void*)(_pipe._buffer + read), (int)sz);
+        }
+
+        public void Advance(uint count)
+        {
+            var read = _pipe._readIdx;
+            var write = _pipe._writeIdx;
+
+            if (read <= write)
+            {
+                if (count > write - read)
+                {
+                    throw new InvalidOperationException();
+                }
+
+                read += count;
+            }
+            else
+            {
+                var sz = Math.Min(count, _pipe.Size - read);
+
+                read += sz;
+                if (read > _pipe.Size - 1)
+                {
+                    read = 0;
+                }
+                count -= sz;
+
+                if (count > 0)
+                {
+                    if (count > write)
+                    {
+                        throw new InvalidOperationException();
+                    }
+
+                    read = count;
+                }
+            }
+
+            if (read == write)
+            {
+                // If the read pointer catches up to the write pointer, then the pipe is empty.
+                // As a performance optimization, set both to 0. This should improve the chances cache lines are hit.
+                _pipe._readIdx = 0;
+                _pipe._writeIdx = 0;
+            }
+            else
+            {
+                _pipe._readIdx = read;
+            }
+        }
+
+        public void Close() => _pipe._closed = true;
+
+        public bool IsClosed => _pipe._closed;
+    }
+
+    private IntPtr _buffer;
+    private readonly uint _bufferSize;
+    private volatile uint _writeIdx;
+    private volatile uint _readIdx;
+    private bool _closed;
+
+    public PipeWriter Writer { get; }
+    public PipeReader Reader { get; }
+
+    public uint Size => _bufferSize;
+
+    public bool Closed => _closed;
+
+    public RingPipe(uint size)
+    {
+        var pageSize = (uint)Environment.SystemPageSize;
+        var adjustedSize = (size + pageSize - 1) & ~(pageSize - 1);
+
+        // If Windows - Do windows thing
```

Comment:
Missing some code?

Reply:
Yes, I am not home yet to add/test the windows part.